libyul/ASTNodeRegistry.cpp

```diff
@@ -0,0 +1,150 @@
+	std::vector<uint8_t> labelVisited (_labels.size(), false);
+	for (auto const& name: _nameToLabelMapping)
+	{
+		yulAssert(name < _labels.size());
+		// it is possible to have multiple references to empty / ghost
+		yulAssert(
+			name < 2 || !labelVisited[name],
+			fmt::format("YulName {} (label \"{}\") is not unique.", name, _labels[name])
+		);
+		labelVisited[name] = true;
+	}
+	m_labels = _labels;
+	m_nameToLabelMapping = _nameToLabelMapping;
+}
+
+size_t ASTNodeRegistry::nameToLabelIndex(NodeId const _name) const
+{
+	yulAssert(_name < m_nameToLabelMapping.size());
+	return m_nameToLabelMapping[_name];
+}
+
+std::string_view ASTNodeRegistry::operator()(NodeId const _name) const
+{
+	auto const labelIndex = nameToLabelIndex(_name);
+	if (labelIndex == ghostId())
+		return lookupGhost(_name);
+	return m_labels[nameToLabelIndex(_name)];
+}
+
+std::optional<ASTNodeRegistry::NodeId> ASTNodeRegistry::findNameForLabel(std::string_view const _label) const {
+	if (_label.empty())
+		return emptyId();
+	if (_label == ghostPlaceholder)
+		return NodeId{1};
+	for (NodeId name = 2; name <= maximumId(); ++name)
+		if ((*this)(name) == _label)
+			return name;
+	return std::nullopt;
+}
+
+std::string_view ASTNodeRegistry::lookupGhost(NodeId const _name) const
+{
+	yulAssert(nameToLabelIndex(_name) == ghostId());
+	auto const [it, _] = m_ghostLabelCache.try_emplace(_name, fmt::format("GHOST[{}]", _name));
+	return it->second;
+}
+
+ASTNodeRegistryBuilder::LabelToNameMapping::LabelToNameMapping():
+	m_mapping{{"", 0}, {ASTNodeRegistry::ghostPlaceholder, 1}}
+{}
+
+std::tuple<ASTNodeRegistry::NodeId, bool> ASTNodeRegistryBuilder::LabelToNameMapping::tryInsert(
+	std::string_view const _label,
+	ASTNodeRegistry::NodeId const _name
+)
+{
+	yulAssert(_label != ASTNodeRegistry::ghostPlaceholder);
+	auto const [it, emplaced] = m_mapping.try_emplace(std::string{_label}, _name);
+	return std::make_tuple(it->second, emplaced);
+}
+
+ASTNodeRegistryBuilder::ASTNodeRegistryBuilder():
+	m_nextNameId(2)
+{}
+
+ASTNodeRegistryBuilder::ASTNodeRegistryBuilder(ASTNodeRegistry const& _existingRegistry)
+{
+	for (size_t i = 2; i <= _existingRegistry.maximumId(); ++i)
+	{
+		auto const existingLabel = _existingRegistry(i);
+		if (!existingLabel.empty())
+		{
+			auto const [_, inserted] = m_mapping.tryInsert(_existingRegistry(i), i);
+			yulAssert(inserted);
+		}
+	}
+	m_nextNameId = _existingRegistry.maximumId() + 1;
+}
+
+ASTNodeRegistry::NodeId ASTNodeRegistryBuilder::define(std::string_view const _label)
+{
+	auto const [nameId, inserted] = m_mapping.tryInsert(_label, m_nextNameId);
+	if (inserted)
+		m_nextNameId++;
+	return nameId;
+}
+
+ASTNodeRegistry ASTNodeRegistryBuilder::build() const
+{
+	auto const& mapping = m_mapping.data();
+	yulAssert(mapping.contains(""));
+	yulAssert(mapping.at("") == 0);
+	yulAssert(mapping.contains(ASTNodeRegistry::ghostPlaceholder));
+	yulAssert(mapping.find(ASTNodeRegistry::ghostPlaceholder)->second == 1);
```

Comment:
Why not `at`?  You assert that it is in the mapping, and you use `at` for empty string.
```suggestion
	yulAssert(mapping.at(ASTNodeRegistry::ghostPlaceholder) == 1);
```

Reply:
Initially I had the `ghostPlaceholder` as `std::string_view`. In that case you cannot use `at` or `operator[]` on a map `std::string -> X`, even with the opaque comparison operator. Only `find` does the trick. But now that it no longer is a `string_view`, I can indeed use `at` :)